src/crates.rs

```diff
@@ -0,0 +1,277 @@
+    pub fn tarball(self, output: &Path) -> Result<Option<DirEntry>, Error> {
+        let license = if crate::must_copy_license(&self.config.package.license) {
+            p("LICENSE file will be installed manually.".bold().yellow());
+            Some(self.license_file()?)
+        } else {
+            None
+        };
+        // Possible refactor target with super::work
+        let env_cargo_target: PathBuf = match std::env::var_os("CARGO_TARGET_DIR") {
+            Some(p) => PathBuf::from(p),
+            None => PathBuf::from("target"),
+        };
+        let cargo_target = self
+            .tempdir_handle
+            .path()
+            .join(self.crate_file_prefix)
+            .join(env_cargo_target);
+        super::tarball(
+            self.musl,
+            cargo_target.as_ref(),
+            output,
+            license.as_ref(),
+            self.config,
+        )?;
+        Ok(license)
+    }
+    /// Get the path to the first file with a name starting with LICENSE in the current directory.
+    fn license_file(&self) -> Result<DirEntry, Error> {
+        crate::license_file(Some(
+            self.tempdir_handle
+                .as_ref()
+                .join(&self.crate_file_prefix)
+                .as_ref(),
+        ))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::CrateFile;
+    use crate::Config;
+    use cargo_aur::Package;
+    use std::path::PathBuf;
+    /// Download a crate and check the sha256sums against the expected value.
+    #[test]
+    fn test_sha256() {
+        let package = Package {
+            name: "cargo-aur".into(),
+            version: "1.7.1".into(),
+            authors: Vec::new(),
+            description: String::new(),
+            repository: String::new(),
+            license: "MIT".into(),
+            metadata: None,
+            homepage: None,
+            documentation: None,
+        };
+        let cfg = Config {
+            package,
+            bin: Vec::new(),
+        };
+        let crate_file = CrateFile::download_new(&cfg).unwrap();
+        let crate_sha = crate_file.get_sha256sum().unwrap();
+        // Hardcoded expected sha
+        let expected_sha = "a475dd0482fb5b2782f7edf1728430fa5c2dabc9298771b60e7e7ae708eab31c";
+        assert_eq!(crate_sha, expected_sha);
+    }
+    /// Test that the package can build correctly.
+    #[test]
+    fn test_build() {
+        let package = Package {
+            name: "cargo-aur".into(),
+            version: "1.7.1".into(),
+            authors: Vec::new(),
+            description: String::new(),
+            repository: String::new(),
+            license: "MIT".into(),
+            metadata: None,
+            homepage: None,
+            documentation: None,
+        };
+        let cfg = Config {
+            package,
+            bin: Vec::new(),
+        };
+        CrateFile::download_new(&cfg)
+            .unwrap()
+            .build(false)
+            .expect("Expected build to succeed");
+    }
```

Comment:
Same here, with anything that involves the network.

Reply:
Actioned